Double-check the code for key/value gradient reduction in the case of MQA, when tensor-model-parallel > 1, and for distributed optim

Looking again at this code: https://github.com/bigcode-project/Megatron-LM/blob/multi-query-attention/megatron/optimizer/optimizer.py#L269
1 - Aren't the gradients of the biases missing from the reduction?

2 - The distributed optimizer misses this reduction: https://github.com/bigcode-project/Megatron-LM/blob/8169dec7a78a84cefd65ad69c5060f2a1fba15a3/megatron/optimizer/distrib_optimizer.py#L522

There is also a crash when combining with sequence parallelism:
```
2022-12-09 04:21:00,367 (worker_1) :   File "/app/megatron/model/transformer.py", line 415, in forward
2022-12-09 04:21:00,367 (worker_1) :     attention_probs = self.scale_mask_softmax(attention_scores,
2022-12-09 04:21:00,367 (worker_1) :   File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1129, in _call_impl
2022-12-09 04:21:00,367 (worker_1) :     return forward_call(*input, **kwargs)
2022-12-09 04:21:00,367 (worker_1) :   File "/app/megatron/model/fused_softmax.py", line 161, in forward
2022-12-09 04:21:00,367 (worker_1) :     return self.forward_fused_softmax(input, mask)
2022-12-09 04:21:00,367 (worker_1) :   File "/app/megatron/model/fused_softmax.py", line 192, in forward_fused_softmax
2022-12-09 04:21:00,367 (worker_1) :     assert sq == sk, "causal mask is only for self attention"
2022-12-09 04:21:00,367 (worker_1) : AssertionError: causal mask is only for self attention
```
I did some investigating and found that the key layer skips the sequence-parallel gather because it's a normal linear layer  instead of the tensor-parallel one, so it ends up with the wrong sequence length.

--- a/megatron/model/transformer.py
+++ b/megatron/model/transformer.py
@@ -608,8 +608,12 @@ class ParallelAttention(MegatronModule):
              key_layer,
              value_layer) = mpu.split_tensor_along_last_dim(mixed_x_layer, 3)
         elif self.attention_type == AttnType.self_attn and self.attention_head_type == 'multiquery':
+            kv_input=hidden_states
+            if get_args().sequence_parallel:
+                # The linear layer doesn't gather the sequence-parallel.
+                kv_input = mpu.gather_from_sequence_parallel_region(kv_input, tensor_parallel_output_grad=False)
             # Attention heads [sq, b, h] --> [sq, b, (2 * hn)]
-            mixed_kv_layer = self.key_value(hidden_states)
+            mixed_kv_layer = self.key_value(kv_input)
 
             # [sq, b, (2 * hn)] --> [sq, b, np (expanded), 2 * hn]
             # new_tensor_shape = mixed_kv_layer.size()[:-1] + \

--- a/megatron/optimizer/distrib_optimizer.py
+++ b/megatron/optimizer/distrib_optimizer.py
@@ -541,6 +541,12 @@ class DistributedOptimizer(MixedPrecisionOptimizer):
         self.allreduce_embedding_grads(args)
         timers('backward-embedding-all-reduce').stop()
 
+        # All-reduce key-value grads if needed.
+        if args.attention_head_type == "multiquery":
+            timers('backward-key-value-all-reduce').start()
+            self.allreduce_key_value_grads(args)
+            timers('backward-key-value-all-reduce').stop()
+
         # Reduce-scatter setup.
         timers('backward-params-all-reduce').start()
         data_parallel_rank = mpu.get_data_parallel_rank()

--- a/megatron/optimizer/optimizer.py
+++ b/megatron/optimizer/optimizer.py
@@ -267,18 +267,28 @@ class MegatronOptimizer(ABC):
         self.allreduce_position_embedding_grads(args)
     
     def allreduce_key_value_grads(self, args):
-        # TODO: models[0] ?
-        unwrapped_model = self.models[0]
-        unwrapped_model = unwrap_model(
-                unwrapped_model, (torchDDP, LocalDDP, Float16Module))
-        for layer in unwrapped_model.language_model.encoder.layers:
-            kv_weight = layer.self_attention.key_value.weight
-            if args.DDP_impl == 'local':
-                grad = kv_weight.main_grad
-            else:
-                grad = kv_weight.grad
-            torch.distributed.all_reduce(grad, group=mpu.get_tensor_model_parallel_group())
-
+        """
+        Reduce the gradients for the key_value weights and biases for multi-query attention.
+        Coalesce the bias grads to avoid too many small reductions,
+        but not the weight grads since it could cause memory issues.
+        """
+        grads=[]
+        for model_module in self.models:
+            unwrapped_model = unwrap_model(
+                    model_module, (torchDDP, LocalDDP, Float16Module))
+            for layer in unwrapped_model.language_model.encoder.layers:
+                kv_weight = layer.self_attention.key_value.weight
+                grad = kv_weight.main_grad if args.DDP_impl == 'local' else kv_weight.grad
+                torch.distributed.all_reduce(grad, group=mpu.get_tensor_model_parallel_group())
+                kv_bias = layer.self_attention.key_value.bias
+                grads.append(kv_bias.main_grad if args.DDP_impl == 'local' else kv_bias.grad)
+        if len(grads)>0:
+            coalesced = _flatten_dense_tensors(grads)
+            torch.distributed.all_reduce(
+                coalesced, group=mpu.get_tensor_model_parallel_group())
+            for buf, synced in zip(grads, _unflatten_dense_tensors(
+                    coalesced, grads)):
+                buf.copy_(synced)
 
     def allreduce_layernorm_grads(self, args):
         """All-reduce layernorm grads (for sequence parallelism)."""

